### PR TITLE
feat: change challenge strategy

### PR DESCRIPTION
**Pull request by bilbeyt**
Closes: #1064

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -83,6 +84,11 @@ ignore_whitespace = true
 module = ["ape.*", "twisted.*"]
 ignore_missing_imports = true
 
+[[tool.mypy.overrides]]
+module = ["beamer.tests.*"]
+disallow_untyped_defs = false
+disable_error_code = "attr-defined,name-defined"
+
```
Comment (istankovic) [inline pyproject.toml:91]: Eh, here we get this again...

Comment (bilbeyt) [reply pyproject.toml]: gone now :D

File: beamer/agent/chain.py [hunk 2]
```diff
@@ -438,10 +453,10 @@ def maybe_challenge(claim: Claim, context: Context) -> bool:
     stake = claim.get_minimum_challenge_stake(initial_claim_stake)
 
     # TODO: have a central variable and proper L1 cost calculations
-    l1_cost = Wei(initial_claim_stake + 10**15)
+    l1_cost = Wei(initial_claim_stake + get_l1_cost(context))
```
Comment (fredo) [inline beamer/agent/chain.py:456]: I think the above TODO comment can be removed

File: beamer/tests/agent/test_challenge.py [hunk 3]
```diff
@@ -93,21 +105,31 @@ def test_challenge_2(request_manager, token, config):
...
+            while (claim := agent.get_context(direction).claims.get(claim_event.claimId)) is None:
+                sleeper.sleep(0.1)
 
-        ape.chain.mine(timestamp=claim.termination)
-        request_manager.withdraw(claim.claimId, sender=agent.address)
+        with Sleeper(5) as sleeper:
+            while (request := agent.get_context(direction).requests.get(request_id)) is None:
+                sleeper.sleep(0.1)
+
+        with freeze_time(
```
Comment (fredo) [inline beamer/tests/agent/test_challenge.py:120]: I assume if we use the chain's timestamp for determining the L1 finalization, we don't need this context manager here, do we?

File: beamer/tests/agent/test_challenge.py [hunk 4]
```diff
@@ -408,3 +482,120 @@ def test_challenge_7(request_manager, fill_manager, token, config):
...
+
+# Scenario:
+#
+# Bob          Charlie
+# --------------------------
+#                  dishonest claim
+# honest claim
+# challenge
+#                  challenge
+# Winner: Charlie
```
Comment (fredo) [inline beamer/tests/agent/test_challenge.py:495]: Why is charlie the winner if he sent a dishonest claim

File: beamer/tests/agent/test_challenge.py [hunk 5]
```diff
@@ -408,3 +482,120 @@ def test_challenge_7(request_manager, fill_manager, token, config):
...
+
+
+# Scenario:
+#
+# Bob          Charlie
+# --------------------------
+# claim
+#                  challenge
+#
+# Winner: Charlie
```
Comment (fredo) [inline beamer/tests/agent/test_challenge.py:560]: also here I assume Bob should be the winner

File: beamer/agent/state_machine.py [hunk 6]
```diff
@@ -154,7 +154,7 @@ def _proof_ready_for_l1_relay(request: Request) -> bool:
     )
 
 
-def _timestamp_is_l1_finalized(
+def timestamp_is_l1_finalized(
```
[the remark this answers is not in the document]
Comment (bilbeyt) [reply beamer/agent/state_machine.py:157]: i think we decided to continue like this

File: beamer/tests/agent/test_challenge.py (hunk 3 above)
Comment (bilbeyt) [reply beamer/tests/agent/test_challenge.py:120]: same here, as we continue like before i will keep this.

File: beamer/tests/agent/test_challenge.py (hunk 5 above)
Comment (bilbeyt) [reply beamer/tests/agent/test_challenge.py]: as other pr had tests for resolution and invalidation i skipped these 2 and created another test to make sure we dont drain agent.

File: beamer/tests/agent/test_challenge.py [hunk 7]
```diff
@@ -140,17 +162,30 @@ def test_challenge_3(request_manager, fill_manager, token, config):
...
         agent.start()
 
-        # Get Bob's challenge.
-        claim = collector.next_event()
-        assert claim is not None
-        assert (
-            claim.challengerStakeTotal > claim.claimerStake
-            and claim.lastChallenger == agent.address
-        )
+        with Sleeper(5) as sleeper:
```
[the remark this answers is not in the document]
Comment (bilbeyt) [reply beamer/tests/agent/test_challenge.py:174]: i need claim invalidation timestamp

File: beamer/agent/chain.py [hunk 8]
```diff
@@ -522,6 +538,13 @@ def _timestamp_is_l1_finalized(
     return int(time.time()) > timestamp + target_chain_finality
 
 
+def get_l1_cost(context: Context) -> int:
```
Comment (fredo) [inline beamer/agent/chain.py:541]: in `_l1_resolution_threshold_reached` there is still a code duplicate of this. Can you please reuse get_l1_cost there?

File: beamer/tests/agent/test_challenge.py [hunk 9]
```diff
@@ -94,22 +107,32 @@ def test_challenge_2(request_manager, token, config):
...
-        ape.chain.mine(timestamp=claim.termination)
-        request_manager.withdraw(claim.claimId, sender=agent.address)
+        with Sleeper(5) as sleeper:
+            while (request := agent.get_context(direction).requests.get(request_id)) is None:
+                sleeper.sleep(0.1)
+
+        with freeze_time(
+            datetime.fromtimestamp(request.fill_timestamp + finality_period + 1), tick=True
+        ):
+            # Bob's claim.
```
Comment (fredo) [inline beamer/tests/agent/test_challenge.py:126]: would also rename it to `Bob's challenge`

File: beamer/tests/agent/test_challenge.py [hunk 10]
```diff
@@ -491,9 +571,99 @@ def test_invalidation(
...
+    new_chain_params = (1, *chain_params[1:])
+    request_manager.updateChain(ape.chain.chain_id, *new_chain_params)
+
+    request_id = make_request(request_manager, token, requester, target, amount)
+
+    with Sleeper(5) as sleeper:
+        while (request := agent.get_context(direction).requests.get(request_id)) is None:
+            sleeper.sleep(0.1)
+
+    with Sleeper(5) as sleeper:
```
Comment (fredo) [inline beamer/tests/agent/test_challenge.py:604]: this check is not necessary as we wait for the claim in the lines below. The claim will only come if the request was seen and claimed

File: beamer/tests/agent/test_challenge.py [hunk 11]
```diff
@@ -491,9 +571,99 @@ def test_invalidation(
...
+        assert challenge_claim.claimerStake > l1_cost
+
+        request_manager.challengeClaim(
+            claim.id, sender=exploiter, value=challenge_claim.claimerStake + 1
+        )
+
+        exploiter_claim = collector.next_event()
+        assert exploiter_claim is not None
+
+        # l1 cost is ready no more challenge
```
Comment (fredo) [inline beamer/tests/agent/test_challenge.py:649]: `l1 cost are staked by exploiter, so don't challenge again`

**Comment by fredo** [comment]
left a view comments otherwise looks good!

File: beamer/agent/chain.py [hunk 12]
```diff
@@ -412,10 +414,25 @@ def maybe_challenge(claim: Claim, context: Context) -> bool:
...
+            request.fill_timestamp, context, context.target_chain_id
+        )
+    elif claim.invalidation_timestamp is not None:
+        finalized = _timestamp_is_l1_finalized(
+            claim.invalidation_timestamp, context, context.target_chain_id
+        )
+    if not finalized:
+        return False
+
+    if _l1_resolution_threshold_reached(claim, context):
```
[the remark this answers is not in the document]
Comment (bilbeyt) [reply beamer/agent/chain.py:430]: maybe i understood you wrongly before, i was thinking like this but then remembered like you also wanted to put this here. just removing it now.

File: beamer/tests/agent/test_challenge.py [hunk 13]
```diff
@@ -491,9 +571,99 @@ def test_invalidation(
         while agent.get_context(direction).l1_resolutions.get(claim.invalidation_tx) is None:
             sleeper.sleep(0.1)
 
+    ape.chain.provider._make_request(
+        "evm_setAccountBalance",
+        [l1_messenger.address, encode_hex(str(ape.convert("1 ether", int)))],
+    )
+
```
Comment (istankovic) [inline beamer/tests/agent/test_challenge.py:578]: Can we avoid using the private `_make_request` function here?

Comment (bilbeyt) [reply beamer/tests/agent/test_challenge.py:571]: oo sorry, actually l1_messenger is contract that has no payable so i cant randomly send funds to it with test account

**Pull request merged**